File: aizu/Introduction-to-Algorithms-and-Data-Structures/ALDS1_5_B_MergeSort/ALDS1_5_B.js
"use strict"

let cnt = 0
const mergeSort = (xs, left, right) => {
  if (left + 1 < right) {
    const mid = Math.floor((left + right) / 2)
    mergeSort(xs, left, mid)
    mergeSort(xs, mid, right)
    merge(xs, left, mid, right)
  }
}
const merge = (xs, left, mid, right) => {
  const n1 = mid - left
  const n2 = right - mid
  const ls = []
  for (let i = 0; i < n1; i++) {
    ls[i] = xs[left + i]
  }
  ls.push(1000000000)
  const rs = []
  for (let i = 0; i < n2; i++) {
    rs[i] = xs[mid + i]
  }
  rs.push(1000000000)
  let i = 0
  let j = 0
  for (let k = left; k < right; k++) {
    cnt++
    if (ls[i] < rs[j]) {
      xs[k] = ls[i]
      i++
    } else {
      xs[k] = rs[j]
      j++
    }
  }
}

const solver = (n, xs) => {
  mergeSort(xs, 0, n)
  return xs
}

let rows = 0
let n = 0
const reader = require("readline").createInterface({
  input: process.stdin,
  output: process.stdout,
})
reader.on("line", (line) => {
  rows++
  if (rows === 1) {
    n = Number(line.trim())
  } else {
    const xs = line.trim().split(" ").map(x => Number(x))
    const result = solver(n, xs)
    console.log(result.reduce((acc, x) => `${acc} ${x}`))
    console.log(cnt)
    process.exit(0)
  }
})

File: aizu/Introduction-to-Algorithms-and-Data-Structures/ALDS1_5_B_MergeSort/ALDS1_5_B.rb
$n = 0

def merge(sorted_left, sorted_right)
  arr = []
  l_p = r_p = 0
  max = 1000000000
  sorted_left.push(max)
  sorted_right.push(max)
  l_v = r_v = 0
  while l_v != max || r_v != max
    flag = true
    l_v = sorted_left[l_p].to_i
    while flag
      r_v = sorted_right[r_p].to_i
      if l_v == max && r_v == max then
        return arr
      elsif l_v < r_v then
        arr.push(l_v)
        l_p = l_p + 1
        flag = false
      else
        arr.push(r_v)
        r_p = r_p + 1
      end
      $n = $n + 1
    end
  end
  arr
end

def mergeSort(arr)
  if arr.size == 0 then
    arr
  elsif arr.size == 1 then
    arr
  else
    m = arr.size / 2
    left = arr.take(m)
    right = arr.drop(m)
    merge(mergeSort(left), mergeSort(right))
  end
end

n = STDIN.gets.to_i
arr = STDIN.gets.split(' ')
puts mergeSort(arr).join(' ')
puts $n

